Update `get_axis_coord()` to interpret more keys

tomvothecoder opened the pull request:
## Description
<!--
  Please include a summary of the change and which issue is fixed.
  Please also include relevant motivation and context.
  List any dependencies that are required for this change.
-->

Summary of Changes
- Closes #260 
- Closes #215 
- Update `get_axis_coord()` to interpret more keys
   - Rename `get_coord_var()` to `get_axis_coord()`
   - Update `add_missing_bounds()`, `add_bounds()`, and `get_bounds()` logic to use updated `get_axis_coord()`
   - Replace `GENERIC_AXIS_MAP` with `CF_NAME_MAP`
   - Add `CFAxisName`, `CFStandardName`, and `ShortName` type aliases
   - Update variable names in `_add_bounds()` for clarity
   - Move `add_bounds` kwarg in `open_dataset()` and `open_mfdataset()` to a higher position due to importance
- Update regridder APIs to use `get_axis_coord()`
   - Update references to `.cf.get_bounds()` to `.bounds.get_bounds()`
   - Update `grid.create_grid()` to set the `"bounds"` attr on the coord var to avoid `add_missing_bounds()` to attempt to add bounds for global mean grids
- Add `get_axis_dim()`
   - Update spatial averaging APIs to use this function
   - Update `_align_lon_bounds_to_360` to use this function
- Misc Refactor
   - Update name of attr `self._dim_name` to `self._dim` in `TemporalAccessor`
   - Update "is_generated" attr to "xcdat_bounds"
   - Add handles for lat ` "units" ` attr in `_add_bounds()`

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

If applicable:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass with my changes (locally and CI/CD build)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

tomvothecoder commented on xcdat/axis.py:
```diff
@@ -2,32 +2,91 @@
+        The CF-compliant axis name.
+
+    Returns
+    -------
+    xr.DataArray
+        The coordinate variable.
+
+    Raises
+    ------
+    KeyError
+        If the coordinate variable was not found in the dataset.
+
+    Notes
+    -----
+    Refer to [1]_ for a list of CF-compliant "axis" and "standard_name" attr
+    names that can be interpreted by ``cf_xarray``.
+
+    References
+    ----------
+
+    .. [1] https://cf-xarray.readthedocs.io/en/latest/coord_axes.html#axes-and-coordinates
+    """
+    keys = CF_NAME_MAP[axis]
+    coord_var = None
+
+    for key in keys:
+        try:
+            coord_var = dataset.cf[key]
+            break
+        except KeyError:
+            pass
+
+    if coord_var is None:
+        raise KeyError(
+            f"A coordinate variable for the {axis} axis was not found in the dataset. "
+            "Make sure the coordinate variable exists in the dataset and either the "
+            "'axis' or 'standard_name' attr is set, or the name of the dim and "
+            "coords is the common short name (e.g.,'lat')."
+        )
+    return coord_var
```
Function to be reviewed

tomvothecoder commented on xcdat/axis.py:
```diff
@@ -2,32 +2,91 @@
-    "X": "X",
-    "time": "T",
-    "T": "T",
+# https://cf-xarray.readthedocs.io/en/latest/coord_axes.html#axis-names
+CFAxisName = Literal["X", "Y", "T"]
+# https://cf-xarray.readthedocs.io/en/latest/coord_axes.html#coordinate-names
+CFStandardName = Literal["latitude", "longitude", "time"]
+ShortName = Literal["lat", "lon"]
+
+# The key is the accepted value for method and function arguments, and the
+# values are the CF-compliant axis and standard names that are interpreted in
+# the dataset.
+CF_NAME_MAP: Dict[CFAxisName, List[Union[CFAxisName, CFStandardName, ShortName]]] = {
+    "X": ["X", "longitude", "lon"],
+    "Y": ["Y", "latitude", "lat"],
+    "T": ["T", "time"],
```
The new axis coordinate mapping table structure

tomvothecoder commented on tests/test_axis.py:
```diff
@@ -280,3 +280,56 @@ def test_swap_from_180_to_360_and_sorts_with_prime_meridian_cell_in_lon_bnds(sel
+
+        assert lat.identical(ds.lat)
+
+    def test_returns_coord_var_if_non_cf_name_is_set(self):
```
```suggestion
    def test_returns_coord_var_if_short_name_is_set(self):
```

tomvothecoder commented on tests/test_bounds.py:
```diff
@@ -53,55 +53,69 @@ def setup(self):
 class TestGetBounds:
     @pytest.fixture(autouse=True)
     def setup(self):
+
```
```suggestion
```

tomvothecoder commented on xcdat/bounds.py:
```diff
@@ -120,9 +117,8 @@ def keys(self) -> List[str]:
-        (longitude), and Y (latitude).
+        This function loops through the Dataset's axes and adds coordinate
+        bounds for an axis that doesn't have any.
 
```
**TODO: Add docstring for exceptions**

tomvothecoder commented on xcdat/bounds.py:
```diff
@@ -204,32 +212,35 @@ def add_bounds(self, axis: BoundsAxis, width: float = 0.5) -> xr.Dataset:
         ------
         ValueError
             If bounds already exist. They must be dropped first.
```
TODO: Add docstring for ValueError

tomvothecoder commented on xcdat/bounds.py:
```diff
@@ -204,32 +212,35 @@ def add_bounds(self, axis: BoundsAxis, width: float = 0.5) -> xr.Dataset:
             self.get_bounds(axis)
             raise ValueError(
-                f"{axis} bounds already exist. Drop them first to add new bounds."
+                f"'{axis}' bounds already exist. Drop them first to add new bounds."
```
```suggestion
                f"{axis} bounds already exist. Drop them first to add new bounds."
```

tomvothecoder commented on xcdat/bounds.py:
```diff
@@ -134,28 +130,30 @@ def add_missing_bounds(self, width: float = 0.5) -> xr.Dataset:
         ----------
-        axis : BoundsAxis
-            The axis key.
+        axis : CFAxisKey
```
```suggestion
        axis : CFAxisName
```

tomvothecoder commented on xcdat/bounds.py:
```diff
@@ -167,30 +165,40 @@ def get_bounds(self, axis: BoundsAxis) -> xr.DataArray:
 
         Parameters
         ----------
         axis : BoundsAxis
```
```suggestion
        axis : CFAxisName
```

tomvothecoder commented on README.rst:
```diff
@@ -89,8 +89,8 @@ Things we are striving for:
 
-  * Coordinates and bounds are retrieved with ``cf_xarray`` using the standard ``axis`` and ``coordinate`` name attributes
-  * Ability to operate on both longitudinal axis orientations, [0, 360) and [-180, 180)
+   * Coordinates and bounds are retrieved with ``cf_xarray`` using either the ``standard_name`` attr, the `axis` attr, or the dimension short name (e.g., 'lat' for latitude).
+   * Ability to operate on both longitudinal axis orientations, [0, 360) and [-180, 180)
```
```suggestion
  * Coordinates and bounds are retrieved with ``cf_xarray`` using either the ``standard_name`` attr, the `axis` attr, or the dimension short name (e.g., 'lat' for latitude).
  * Ability to operate on both longitudinal axis orientations, [0, 360) and [-180, 180)
```

tomvothecoder replied:
```suggestion
  * Coordinates and bounds are retrieved with ``cf_xarray`` using either the ``standard_name`` attr, the ``axis`` attr, or the dimension short name (e.g., 'lat' for latitude).
  * Ability to operate on both longitudinal axis orientations, [0, 360) and [-180, 180)
```

tomvothecoder commented on docs/index.rst:
```diff
@@ -50,7 +50,7 @@ Things we are striving for:
 * Robust handling of coordinates and its associated bounds
 
-  * Coordinates and bounds are retrieved with ``cf_xarray`` using the standard ``axis`` and ``coordinate`` name attributes
+  * Coordinates and bounds are retrieved with ``cf_xarray`` using either the ``standard_name`` attr, the `axis` attr, or the dimension short name (e.g., 'lat' for latitude).
```
```suggestion
  * Coordinates and bounds are retrieved with ``cf_xarray`` using either the ``standard_name`` attr, the ``axis`` attr, or the dimension short name (e.g., 'lat' for latitude).
```

pochedls commented:
I was first seeing if this PR resolves the code snippet I outlined [here](https://github.com/xCDAT/xcdat/issues/215#issuecomment-1162284773), but I am getting the same error in `spatial._validate_axis_arg`. Am I doing something wrong or should this PR not resolve that specific issue?

tomvothecoder commented:
Thanks @pochedls, I failed to update the spatial averaging methods to agnostically interpret the dimension names for the spatial axes.

Here's the commit that fixes #215: https://github.com/xCDAT/xcdat/pull/262/commits/436f29b5b4039ac4c7cfcd117f7b4da9cf4b6ce3

I tested it using the code snippet from that PR it is now working.
```
import xcdat

fn = "/p/user_pub/climate_work/pochedley1/surface/gistemp1200_GHCNv4_ERSSTv5.nc"
ds = xcdat.open_dataset(fn)
dsa = ds.spatial.average(data_var="tempanomaly")

print(dsa)

<xarray.Dataset>
Dimensions:      (lat: 90, lon: 180, time: 1706, nv: 2, bnds: 2)
Coordinates:
  * lat          (lat) float32 -89.0 -87.0 -85.0 -83.0 ... 83.0 85.0 87.0 89.0
  * lon          (lon) float32 -179.0 -177.0 -175.0 -173.0 ... 175.0 177.0 179.0
  * time         (time) datetime64[ns] 1880-01-15 1880-02-15 ... 2022-02-15
Dimensions without coordinates: nv, bnds
Data variables:
    time_bnds    (time, nv) datetime64[ns] ...
    tempanomaly  (time) float32 -0.1809 -0.265 -0.09773 ... 0.8567 0.9067 0.8949
    lon_bnds     (lon, bnds) float32 -180.0 -178.0 -178.0 ... 178.0 178.0 180.0
    lat_bnds     (lat, bnds) float32 -90.0 -88.0 -88.0 -86.0 ... 88.0 88.0 90.0
Attributes:
    title:        GISTEMP Surface Temperature Analysis
    institution:  NASA Goddard Institute for Space Studies
    source:       http://data.giss.nasa.gov/gistemp/
    Conventions:  CF-1.6
    history:      Created 2022-03-11 09:08:31 by SBBX_to_nc 2.0 - ILAND=1200,...
```

tomvothecoder commented on xcdat/spatial.py:
```diff
@@ -738,7 +731,17 @@ def _averager(self, data_var: xr.DataArray, axis: List[SpatialAxis]):
         """
         weights = self._weights.fillna(0)
 
+        # The axis's dimension name is extracted directly from its related
+        # coordinate variable. This is is done in case the `axis` or
+        # `standard_name` attr is not set on the coord var for `cf_xarray` to
+        # interpret. For example, if `axis=["Y"]`, the dim name "lat" is
+        # extracted from `self._Y` (a DataArray coord var).
+        dim_names = []
+        for key in axis:
+            dim_name = getattr(self, f"_{key}").name
+            dim_names.append(dim_name)
+
         with xr.set_options(keep_attrs=True):
-            weighted_mean = data_var.cf.weighted(weights).mean(axis)
+            weighted_mean = data_var.cf.weighted(weights).mean(dim=dim_names)
```
`cf_xarray` was attempting to interpret the dimension names using the `"axis"` attr on the coordinate variables. However, this breaks if the `"axis"` attr is not set.

Instead, we get the dimension names directly from the coordinate variable.

pochedls commented on xcdat/spatial.py:
```diff
@@ -301,14 +305,10 @@ def _validate_axis_arg(self, axis: List[SpatialAxis]):
+            coord_var = get_coord_var(self._dataset, key)
+            # Set the attribute for the axis to the coordinate variable if the
+            # coordinate variable exists.
+            self.__setattr__(f"_{key}", coord_var)
```
@tomvothecoder - I didn't understand why you were setting an attribute (e.g., `_lat`). I noticed this because when I tried to call `xcdat.spatial._averager()` directly it gave me an error 

> AttributeError: 'Dataset' object has no attribute '_lat'

Because I had not yet ran `_validate_axis_arg()` (which sets these attributes). I'm sure there is a good reason for this...is there a **simple** explanation for setting these underscore attributes?

pochedls replied:
This makes sense. One consequence is that I cannot call `_averager()` directly even if I have the inputs outlined in the documentation. I think that is probably okay because this is not intended to be a public method.

pochedls replied:
Continuing to work on this. I'm now wondering if similar changes need to be made to the `_validate_axis_arg` in `bounds.py`? I'm now getting a KeyError in `axis.py` at this line.

```
    bounds = xr.concat(
        (bounds, bounds[p_meridian_index, :]), dim=bounds.cf.axes["X"][0]
    )
```

I get an empty dictionary for this dim argument: 

```
print(bounds.cf.axes)
```

> {}

tomvothecoder replied:
I ended up removing the implementation of the underscore attributes because we only need the dimension name(s) for the `axis` arg a few times. It also makes the code much easier to read compared to  using `__setattr__` and `getattr` for the underscore attributes.

I implemented `get_axis_dim()` for this specific metadata:
https://github.com/xCDAT/xcdat/blob/93a4a7fff990bbac9d0bafb57a472c7950e95ae5/xcdat/axis.py#L86-L99

Now in `_average()`:
https://github.com/xCDAT/xcdat/blob/93a4a7fff990bbac9d0bafb57a472c7950e95ae5/xcdat/spatial.py#L723-L732

You should be able to test `_average()` without needing to set the underscore attributes.

I also fixed `bounds._validate_axis_arg()`:
https://github.com/xCDAT/xcdat/blob/93a4a7fff990bbac9d0bafb57a472c7950e95ae5/xcdat/axis.py#L344-L348

tomvothecoder replied:
Marking this as resolved.

durack1 commented:
@tomvothecoder a noob question, if I just checkout this branch, should that allow me to start testing against these edge case files that I've encountered? (or do I have to checkout more than this #262 branch alone)

tomvothecoder commented:
> @tomvothecoder a noob question, if I just checkout this branch, should that allow me to start testing against these edge case files that I've encountered? (or do I have to checkout more than this #262 branch alone)

Exactly, you just need to setup the Python development environment for this repository and you can test the branch in a Jupyter Notebook or Python script/console.

Guide to get that setup: https://xcdat.readthedocs.io/en/latest/contributing.html#local-development

durack1 commented:
> Exactly, you just need to setup the Python development environment

Done that, and checked out this branch #262, but I still get an error so am wondering if I also need to pull the changes from the https://github.com/xCDAT/xcdat/commit/436f29b5b4039ac4c7cfcd117f7b4da9cf4b6ce3 commit?

tomvothecoder commented:
@durack1 You might be behind on commits if you are still seeing errors that were resolved in previous comments, so I think pulling the latest commits might fix them.

durack1 commented:
@tomvothecoder looks like I am up-to-date:
```
(xcdat_dev) bash-4.2$ git pull
FIPS mode initialized
Updating f9f43f7..7bfdf56
```
And I am still hitting a similar issue
```python
(xcdat_dev) bash-4.2$ python
Python 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:25:59) 
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import xcdat
>>> print(xcdat.__version__)
0.2.0
>>> f = "/p/css03/esgf_publish/CMIP6/PMIP/IPSL/IPSL-CM6A-LR/midPliocene-eoi400/r1i1p1f1/AERmonZ/o3/grz/v20190118/o3_AERmonZ_IPSL-CM6A-LR_midPliocene-eoi400_r1i1p1f1_grz_185001-204912.nc"
>>> from xcdat import open_dataset as od
>>> xH = od(f)
Traceback (most recent call last):
  File "~/git/xcdat/xcdat/bounds.py", line 178, in get_bounds
    bounds_key = coord_var.attrs["bounds"]
KeyError: 'bounds'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "~/git/xcdat/xcdat/bounds.py", line 145, in add_missing_bounds
    self.get_bounds(axis)
  File "~/git/xcdat/xcdat/bounds.py", line 180, in get_bounds
    raise KeyError(
KeyError: "The coordinate variable lat has no 'bounds' attr. Make sure the 'bounds' attr is set and the bounds data var exists"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~/git/xcdat/xcdat/dataset.py", line 93, in open_dataset
    ds = _postprocess_dataset(ds, data_var, center_times, add_bounds, lon_orient)
  File "~/git/xcdat/xcdat/dataset.py", line 466, in _postprocess_dataset
    dataset = dataset.bounds.add_missing_bounds()
  File "~/git/xcdat/xcdat/bounds.py", line 147, in add_missing_bounds
    self._dataset = self._add_bounds(axis, width)
  File "~/git/xcdat/xcdat/bounds.py", line 307, in _add_bounds
    and "degree" in coord_var.attrs["units"]
KeyError: 'units'
```

tomvothecoder commented:
```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~/git/xcdat/xcdat/dataset.py", line 93, in open_dataset
    ds = _postprocess_dataset(ds, data_var, center_times, add_bounds, lon_orient)
  File "~/git/xcdat/xcdat/dataset.py", line 466, in _postprocess_dataset
    dataset = dataset.bounds.add_missing_bounds()
  File "~/git/xcdat/xcdat/bounds.py", line 147, in add_missing_bounds
    self._dataset = self._add_bounds(axis, width)
  File "~/git/xcdat/xcdat/bounds.py", line 307, in _add_bounds
    and "degree" in coord_var.attrs["units"]
KeyError: 'units'
```
@durack1  What is happening with that specific dataset is:

1. Attempt to open the dataset (`add_bounds` defaults to `True` to add missing bounds)
2. Lat coordinate variable is missing a `"units"` attr
3. Attempt to clip the newly generated lat bounds in `_add_bounds()`
4. Breaks because the `"units"` attr is missing

Related code:
https://github.com/xCDAT/xcdat/blob/c678c8bfc61c4edda006745c9c4dc535f8c68136/xcdat/bounds.py#L305-L311

Since the `"units"` is usually `"degree"`, I removed that part of the conditional (`            and "degree" in da_coord.attrs["units"]`).

I pushed a commit, so we should be good to go for this specific issue.

tomvothecoder commented:
To follow up on the previous comment, I am working on handling when the lat `"units"` attr is not set or is something other than degrees.

Prototype code:
```
        # Clip latitude bounds at (-90, 90)
        if coord_var.name in ("lat", "latitude", "grid_latitude"):
            units = coord_var.get("units")

            if units is None:
                ds[coord_var.name].attrs["units"] = "degree"
                logger.warning(
                    f"The '{coord_var.name}' coordinate variable is missing "
                    "a 'units' attribute. Assuming 'units' is 'degree'."
                )
            elif "degree" not in units.lower():
                raise ValueError(
                    "The {coord_var.name} coord variable has a 'units' attribute that "
                    "is not in degrees."
                )

            if (coord_var >= -90).all() and (coord_var <= 90).all():
                np.clip(bounds, -90, 90, out=bounds)
```

durack1 commented:
@tomvothecoder great, so [c678c8b](https://github.com/xCDAT/xcdat/commit/c678c8bfc61c4edda006745c9c4dc535f8c68136) doesn't seem to have been added to this branch, right?

And while we are finding edge cases, here are 3 others. I can open these successfully using cdms2 but hit a problem with `xcdat/xcdat/dataset.py, line 84, in open_dataset`
```python
/p/css03/esgf_publish/CMIP6/CMIP/MOHC/HadGEM3-GC31-MM/historical/r2i1p1f3/SIday/sitemptop/gr/v20191218/sitemptop_SIday_HadGEM3-GC31-MM_historical_r2i1p1f3_gr_19900101-19941230.nc
/p/css03/esgf_publish/CMIP6/PAMIP/CCCma/CanESM5/pdSST-futBKSeasSIC/r41i1p2f1/Amon/ua/gn/v20190429/ua_Amon_CanESM5_pdSST-futBKSeasSIC_r41i1p2f1_gn_200004-200105.nc
/p/css03/esgf_publish/CMIP6/ScenarioMIP/EC-Earth-Consortium/EC-Earth3/ssp245/r3i1p1f1/SImon/siu/gn/v20210517/siu_SImon_EC-Earth3_ssp245_r3i1p1f1_gn_203301-203312.nc
```

And this single file is garbled netcdf that has no data, but does have a valid header, so a very edgey case indeed (`cdms2` also has a problem; `xcdat/xcdat/dataset.py", line 90, in open_dataset`):
```python
/p/css03/esgf_publish/CMIP6/FAFMIP/MPI-M/MPI-ESM1-2-LR/faf-passiveheat/r1i1p1f1/Omon/epcalc100/gn/v20210901/epcalc100_Omon_MPI-ESM1-2-LR_faf-passiveheat_r1i1p1f1_gn_187001-188912.nc
```

pochedls commented:
> And while we are finding edge cases, here are 3 others.

@durack1 - I'll let @tomvothecoder decide what is best, but I think based on our recent experience documenting issues with the regridder, I would suggest opening new tickets for these examples unless they are the same underlying issue. 

This will help in diagnosing / resolving the issues quickly and ensuring that the issues are documented in a way that is searchable (and may allow us to understand why changes were made).

durack1 commented:
> unless they are the same underlying issue

@pochedls the 3 files all problems with the time coord, and 2 lines of code in `xcdat/xcdat/dataset.py, line 402-404, in _has_cf_compliant_time`.

The `FAFMIP` edge case could be ignored, but the resulting traceback is not all that helpful in indicating it's an errored file.

Happy to wedge these wherever they are useful, new issues are fine but will be duplicative

tomvothecoder commented:
I just pushed code to handle the `"units"` attr for the lat coordinate variable in `_add_bounds()`:
https://github.com/xCDAT/xcdat/blob/a51e2dd99edaffe00bad8901cb7f4bbc765cdc03/xcdat/bounds.py#L306-L323

---
> > And while we are finding edge cases, here are 3 others.
> 
> @durack1 - I'll let @tomvothecoder decide what is best, but I think based on our recent experience documenting issues with the regridder, I would suggest opening new tickets for these examples unless they are the same underlying issue.
> 
> This will help in diagnosing / resolving the issues quickly and ensuring that the issues are documented in a way that is searchable (and may allow us to understand why changes were made).


> > unless they are the same underlying issue
> 
> @pochedls the 3 files all problems with the time coord, and 2 lines of code in `xcdat/xcdat/dataset.py, line 402-404, in _has_cf_compliant_time`.
> 
> The `FAFMIP` edge case could be ignored, but the resulting traceback is not all that helpful in indicating it's an errored file.
> 
> Happy to wedge these wherever they are useful, new issues are fine but will be duplicative

Thanks @durack1 for documenting these issues. I would normally agree with @pochedls that new opening up new issues would be good so that the scope of the conversation stays related to the PR.  However, it seems like your issues are related to #261, which is a PR independent of this one. I would post your comments over there if that's correct (instead of creating duplicate issues as you mentioned).

---
Also, I think this PR is in good shape after addressing the issues in the previous comments. **I'll do a final review of this PR and merge, unless there is anything other issues you've come across.** This PR and #261 are blockers for v0.3.0, which I'm hoping to release tomorrow (or Monday).

durack1 commented:
@tomvothecoder, @pochedls no problem, the 3 files with time coord issues have been copied across into #261

pochedls commented on xcdat/axis.py:
```diff
@@ -191,9 +268,8 @@ def _align_lon_to_360(dataset: xr.Dataset, p_meridian_index: np.ndarray) -> xr.D
-    lon: xr.DataArray = dataset.bounds._get_coords("lon").copy()
-    lon_name = lon.name
-    lon_bounds: xr.DataArray = dataset.bounds.get_bounds("lon").copy()
+    lon: xr.DataArray = get_axis_coord(ds, "X")
+    lon_bounds: xr.DataArray = dataset.bounds.get_bounds("X")
```
Just making sure it is okay to drop the `.copy()` here.

tomvothecoder replied:
I dropped `.copy()` in this function and meant to add `.copy()` to `get_bounds()`. Thanks for the reminder.

https://github.com/xCDAT/xcdat/blob/51627fd4f9b5caaa2a0d04e6d464b41fa7077446/xcdat/bounds.py#L185-L193

When `.copy()` is and isn't needed:
```
# Coordinate variable
# .copy() not needed, this creates a deep copy
lat = ds.lat
lat.data[0] = None

# Data variable
# .copy() needed, otherwise a shallow copy is created (ref by memory)
lat_bnds = ds.lat_bnds.copy()
lat_bnds.data[0] = None
```

tomvothecoder commented on xcdat/regridder/accessor.py:
```diff
@@ -76,22 +76,15 @@ def grid(self) -> xr.Dataset:
-    def _get_axis_data(
-        self, name: str, standard_name: str
-    ) -> Tuple[xr.DataArray, xr.DataArray]:
-        try:
-            axis = self._ds.cf[name]
-        except KeyError:
-            raise KeyError(
-                f"{standard_name} axis could not be correctly identified in the Dataset"
-            )
```
This logic is already implemented with `get_axis_coord()`, so I chose to reuse it here.

tomvothecoder replied:
This logic is already implemented by `get_coord_var()` so I reused that function here.

The `standard_name` kwarg isn't necessary because `get_axis_coord()` tries to find the coord var using `axis`, `standard_name`, or dim name (e.g., "lat").  Refer to `axis.CF_NAME_MAP`.

tomvothecoder commented on xcdat/regridder/accessor.py:
```diff
@@ -76,22 +76,15 @@ def grid(self) -> xr.Dataset:
 
         try:
-            axis_bnds = self._ds.bounds.get_bounds(axis.name)
+            bounds_var = self._ds.bounds.get_bounds(name)
```
The accepted argument for `bounds.get_bounds()` is now the cf-compliant axis name ("X", "Y", "T", "Z").

You'll notice a bunch of these were updated.

tomvothecoder commented on xcdat/regridder/base.py:
```diff
@@ -32,22 +33,22 @@ def preserve_bounds(
     try:
-        lat_bnds = source_grid.cf.get_bounds("lat")
+        lat_bnds = source_grid.bounds.get_bounds("Y")
     except KeyError:
         pass
     else:
         target[lat_bnds.name] = lat_bnds.copy()
 
     try:
-        lon_bnds = source_grid.cf.get_bounds("lon")
+        lon_bnds = source_grid.bounds.get_bounds("X")
     except KeyError:
         pass
     else:
         target[lon_bnds.name] = lon_bnds.copy()
 
     for dim_name in source.cf.axes:
         try:
-            source_bnds = source.cf.get_bounds(dim_name)
+            source_bnds = source.bounds.get_bounds(dim_name)
```
I replaced the API call to `.cf.get_bounds()` with `bounds.get_bounds()` because our API checks if the coordinate variable exists (e.g., "lat") and its `"bounds"` attr is set to the bounds data var (e.g., "lat_bnds").

If both of those conditions aren't meant, it will raise an error.

tomvothecoder commented on xcdat/regridder/grid.py:
```diff
@@ -376,16 +378,16 @@ def create_global_mean_grid(grid: xr.Dataset) -> xr.Dataset:
+    lat = get_axis_coord(grid, "Y")
     lat_data = np.array([(lat[0] + lat[-1]) / 2.0])
 
-    lat_bnds = grid.cf.get_bounds("lat")
+    lat_bnds = grid.bounds.get_bounds("Y")
     lat_bnds = np.array([[lat_bnds[0, 0], lat_bnds[-1, 1]]])
 
-    lon = grid.cf["lon"]
+    lon = get_axis_coord(grid, "X")
     lon_data = np.array([(lon[0] + lon[-1]) / 2.0])
 
-    lon_bnds = grid.cf.get_bounds("lon")
+    lon_bnds = grid.bounds.get_bounds("X")
```
I replace calls such as `grid.cf["lat"]` with `get_axis_coord()` because sometimes the name of the coordinate variable might not be "lat". 

`get_axis_coord()` tries to find the coord var using `axis`, `standard_name`, or dim name (e.g., "lat").

tomvothecoder commented on xcdat/regridder/grid.py:
```diff
@@ -493,6 +495,7 @@ def create_grid(
         data_vars["lon_bnds"] = lon_bnds
+        lon.attrs["bounds"] = lon_bnds.name
 
     grid = xr.Dataset(data_vars=data_vars, coords={"lat": lat, "lon": lon})
```
The `"bounds"` attr for a coord var must be set to the name of the bounds data var. Otherwise, the subsequent call to `grid.bounds.add_missing_bounds()` will attempt to add bounds even if they already exist because it thinks the bounds don't exist.

If this `"bounds"` attribute is _not_ set, it breaks the test below.

https://github.com/xCDAT/xcdat/blob/e366c9ac6bb1631c251b64d00e733ae0a3919d61/tests/test_regrid.py#L533-L543

Reasons why:

* Bounds already exist but `add_missing_bounds()` doesn't see it because the `"bounds"` attr is not set
* `_add_bounds()` breaks  because there is only 1 coordinate point for `lat` and `lon` (conditional below)

What it should be doing instead:
* Ignore `add_missing_bounds()` altogether

https://github.com/xCDAT/xcdat/blob/e366c9ac6bb1631c251b64d00e733ae0a3919d61/xcdat/bounds.py#L266-L270

tomvothecoder commented on xcdat/bounds.py:
```diff
@@ -144,7 +144,7 @@ def add_missing_bounds(self, width: float = 0.5) -> xr.Dataset:
                     self.get_bounds(axis)
                 except KeyError:
-                    self._dataset = self._add_bounds(axis, width)
+                    self._dataset = self.add_bounds(axis, width)
```
I accidentally referenced the wrong method.

tomvothecoder commented on tests/test_regrid.py:
```diff
@@ -567,12 +567,12 @@ def setup(self):
         ds = fixtures.generate_dataset(True, True)
 
-        ds_no_lat = ds.drop_vars(["lat"])
+        ds_no_lat = ds.drop_dims(["lat"])
 
         with pytest.raises(KeyError):
             ds_no_lat.regridder.grid
 
-        ds_no_lon = ds.drop_vars(["lon"])
```
We have to drop the axis's dimension because its dimension can still exist without coordinates.

The axis should not exist if the corresponding dimension does not exist.

tomvothecoder commented on xcdat/axis.py:
```diff
@@ -21,6 +21,7 @@
     "X": ["X", "longitude", "lon"],
     "Y": ["Y", "latitude", "lat"],
     "T": ["T", "time"],
+    "Z": ["Z", "height", "pressure"],
```
Updated the axis coordinate mapping table

tomvothecoder reviewed (commented):
Hey @jasonb5, I rebased this branch on top of `main` which includes PR #44. Can you review these changes before I merge?

jasonb5 reviewed (approved):
LGTM

tomvothecoder commented on conda-env/readthedocs.yml:
```diff
@@ -15,6 +15,9 @@ dependencies:
     - numpy=1.22.3
     - pandas=1.4.1
     - xarray=2022.3.0
+    - xesmf=0.6.2
+    - esmpy=8.2.0
+    - numba=0.53.1
```
Forgot to add this in #44 so the readthedocs build failed

tomvothecoder commented:
> LGTM

Thanks Jason!
